package ru.practicum.client;

import lombok.extern.slf4j.Slf4j;
import net.devh.boot.grpc.client.inject.GrpcClient;
import org.springframework.stereotype.Service;
import ru.practicum.ewm.stats.proto.InteractionsCountRequestProto;
import ru.practicum.ewm.stats.proto.RecommendationsControllerGrpc;
import ru.practicum.ewm.stats.proto.RecommendedEventProto;

import java.util.Iterator;
import java.util.List;
import java.util.Spliterator;
import java.util.Spliterators;
import java.util.stream.Stream;
import java.util.stream.StreamSupport;

@Service
@Slf4j
public class RecommendationsClient {

    @GrpcClient("analyzer")
    private RecommendationsControllerGrpc.RecommendationsControllerBlockingStub client;

    public Stream<RecommendedEventProto> getRatings(List<Long> ids) {
        log.info("RecommendationClient ratings request");

        try {
            InteractionsCountRequestProto interactionsCountRequestProto = InteractionsCountRequestProto.newBuilder()
                    .addAllEventId(ids)
                    .build();

            Iterator<RecommendedEventProto> iterator = client.getInteractionsCount(interactionsCountRequestProto);
            return asStream(iterator);
        } catch (Exception e) {
            throw new RuntimeException("RecommendationClient ratings request FAILED with message", e);
        }
    }

    private Stream<RecommendedEventProto> asStream(Iterator<RecommendedEventProto> iterator) {
        return StreamSupport.stream(
                Spliterators.spliteratorUnknownSize(iterator, Spliterator.ORDERED),
                false
        );
    }
}
